### PR TITLE
Add codecov

### PR DESCRIPTION
**Pull request by zanebeckwith**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,7 +5,12 @@ before_script:
   - git submodule update --init --recursive
   - mkdir -p build
   - cd build
-  - cmake .. -DCMAKE_BUILD_TYPE=Release
+  - cmake .. -DCMAKE_BUILD_TYPE=Debug
 script:
   - cmake --build .
   - ctest -V
+after_success:
```
Comment (drbild) [inline .travis.yml]: Consider putting these commands in a shell script `.travis/upload-gcov-results` and running as 

```
after_success:
  - .travis/upload-gcov-results
```

Keeps `.travis.yml` organized when there are multiple `after_success` tasks.

File: CMakeLists.txt [hunk 2]
```diff
@@ -23,6 +23,10 @@ project(ecdaa
 
 add_compile_options(-Werror -Wall -Wextra -std=c99)
 
+if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+        SET(CMAKE_C_FLAGS "-g -O0 -fprofile-arcs -ftest-coverage")
```
Comment (drbild) [inline CMakeLists.txt]: Should these be added to a new build type, so that the debug build doesn't always produce the gcov files?  E.g., `DebugWithCoverage`.

See https://cmake.org/Wiki/CMake_FAQ#How_can_I_extend_the_build_modes_with_a_custom_made_one_.3F and https://stackoverflow.com/questions/11437692/how-to-add-a-custom-build-type-to-cmake-targeting-make

Comment (drbild) [inline CMakeLists.txt]: Use `add_compile_options(...)` like line 24, rather than `SET(CMAKE_C_FLAGS ...`.

File: .travis.yml [hunk 3]
```diff
@@ -3,13 +3,19 @@ compiler:
...
   - git submodule update --init --recursive
-  - mkdir -p build
-  - cd build
-  - cmake .. -DCMAKE_BUILD_TYPE=Debug
+  - mkdir -p build/debug
+  - mkdir -p build/release
+  - cd build/debug
+  - cmake ../.. -DCMAKE_BUILD_TYPE=Debug
+  - cd ../release
+  - cmake ../.. -DCMAKE_BUILD_TYPE=Release
```
Comment (drbild) [inline .travis.yml]: End `before_script` with `cd ..`.  Then start `script` with `cd ./release`.  This minimizes state sharing between tasks. Otherwise `script` will break if the `before_script` is changed to finish in a different directory.

File: .travis.yml [hunk 4]
```diff
@@ -3,13 +3,19 @@ compiler:
...
 script:
   - cmake --build .
   - ctest -V
```
Comment (drbild) [inline .travis.yml]: Maybe build and test both `release` and `debug` (or `DebugWithCoverage`) in `script`, not `after_success`.  If the `debug` build or tests fail, that's not a a successful build.  `after_success` should just do the upload of the coverage outputs.

File: CMakeLists.txt (hunk 2 above)
Comment (drbild) [reply CMakeLists.txt]: My though here is that developers may run the debug build locally, but not want the code coverage overhead and output.

File: .travis.yml (hunk 3 above)
Comment (zanebeckwith) [reply .travis.yml]: Agreed. Good call.

File: .travis.yml (hunk 4 above)
Comment (zanebeckwith) [reply .travis.yml]: I wasn't sure what the usual etiquette is. But, I agree with this.

The build-type will now be `DebugWithCoverage`, and just `Debug` is left alone. Good call.

File: CMakeLists.txt (hunk 2 above)
Comment (zanebeckwith) [reply CMakeLists.txt]: The build-type will now be `DebugWithCoverage`, and just `Debug` is left alone. Good call.

**Comment by zanebeckwith** [comment]
I rebased, after accepting (almost all of) your changes. Now, it looks like those comments are hidden. I'm not sure how to "accept" them, or somehow mark them as changed.

File: CMakeLists.txt (hunk 2 above)
Comment (drbild) [reply CMakeLists.txt]: Ok, reason 1 makes sense.  Reason 2 doesn't. I think the semantics are the same.  Each subdirectory gets it's own copy of `CMAKE_C_FLAGS` initialized to the value from its parent directory. (Could be wrong on this.)

But reason 1 wins out regardless.  And yes, I'd switch both to `SET` to be consistent.

In that case you want 
```
SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O0 -fprofile-arcs -ftest-coverage")
```

Otherwise, you're overriding the default build flags and any that the user specified on the command line (see http://voices.canonical.com/jussi.pakkanen/2013/03/26/a-list-of-common-cmake-antipatterns/).  `-g` is implied by the Debug variant, so no need to repeated that.

The whole construct could be shortened from
```
if (CMAKE_BUILD_TYPE STREQUAL "Debug")
    SET(CMAKE_C_FLAGS ...)
endif()
```

to 

```
SET(CMAKE_C_FLAGS_DEBUG ...)
```

as well.

**Comment by drbild** [comment]
You can click on "Show outdated" above to see each change and my comments.

I responded to your `set_compile_options` vs `SET(CMAKE_C_FLAGS ...)` response.

File: CMakeLists.txt (hunk 2 above)
Comment (zanebeckwith) [reply CMakeLists.txt]: As for the differences between the two options, I was going off of https://stackoverflow.com/questions/39501481/difference-between-add-compile-options-and-setcmake-cxx-flags

Yes, good call, I was clobbering the regular flags, good catch.

As for the suffix'd flags variable: that's cool, I didn't know about that!

tl;dr Agree. I've updated to use only `SET(CMAKE_C_FLAGS...`, with the `_DEBUGWITHCOVERAGE` variant used to add the gcov stuff.

File: CMakeLists.txt [hunk 5]
```diff
@@ -21,11 +21,9 @@ project(ecdaa
         LANGUAGES C
         VERSION "0.1.1")
 
-add_compile_options(-Werror -Wall -Wextra -std=c99)
+SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Werror -Wall -Wextra -std=c99")
 
-if (CMAKE_BUILD_TYPE STREQUAL "DebugWithCoverage")
-        SET(CMAKE_C_FLAGS "-g -O0 -fprofile-arcs -ftest-coverage")
-endif()
+SET(CMAKE_C_FLAGS_DEBUGWITHCOVERAGE "${CMAKE_C_FLAGS} -O0 -fprofile-arcs -ftest-coverage")
```
Comment (drbild) [inline CMakeLists.txt]:
```
SET(CMAKE_C_FLAGS_DEBUGWITHCOVERAGE "${CMAKE_C_FLAGS} -O0 -fprofile-arcs -ftest-coverage")
```
to
```
SET(CMAKE_C_FLAGS_DEBUGWITHCOVERAGE "${CMAKE_C_FLAGS_DEBUGWITHCOVERAGE} -O0 -fprofile-arcs -ftest-coverage")
```

Comment (zanebeckwith) [reply CMakeLists.txt]: Does the `SET(CMAKE_C_FLAGS ...` above that just get pulled into every `CMAKE_C_FLAGS*` ?

Comment (drbild) [reply CMakeLists.txt]: I believe so.

Comment (zanebeckwith) [reply CMakeLists.txt]: Gotcha.

I made that change, and the build looks fine.

Comment (drbild) [reply CMakeLists.txt]: From the docs, "For example, if CMAKE_BUILD_TYPE is Debug, then CMAKE_C_FLAGS_DEBUG will be added to the CMAKE_C_FLAGS."

**Comment by drbild** [comment]
LGTM.
